build_test/lib/src/debug_test_builder.dart

```diff
@@ -5,45 +5,123 @@
+    document.querySelector('script[src="packages/test/dart.js"]')?.remove();
+
+    return document.outerHtml;
+  }
+
+  static String _createDebugHtml(String jsScriptPath) => ''
+      '<html>\n'
+      '  <head>\n'
+      '    <script src="$jsScriptPath"></script>\n'
+      '  </head>\n'
+      '</html>\n';
+
+  const DebugTestBuilder();
+
   @override
   final buildExtensions = const {
-    '_test.dart': const ['_test.debug.html']
+    _inputExtension: const [_outputExtension],
   };
 
   @override
-  Future build(BuildStep buildStep) async {
-    var customHtmlId = buildStep.inputId.changeExtension('.html');
-    String debugHtml;
-    var jsScriptPath = '${p.basename(buildStep.inputId.path)}.js';
-    if (await buildStep.canRead(customHtmlId)) {
-      var htmlContent = await buildStep.readAsString(customHtmlId);
-      var document = parse(htmlContent);
+  Future<void> build(BuildStep buildStep) {
+    return _generateDebugHtml(buildStep, buildStep.inputId);
+  }
+}
 
-      // Replace the <link rel="x-dart-test" href="some_path"> tag with
-      var linkTag = document.querySelector('link[rel="x-dart-test"]');
-      var scriptTag = new Element.tag('script');
-      scriptTag.attributes['src'] = jsScriptPath;
-      linkTag.replaceWith(scriptTag);
+/// Generates `text/index.html`, useful for navigating and running tests.
+class DebugIndexBuilder implements Builder {
+  static final _allTests = new Glob(p.join('test', '**$_inputExtension'));
 
-      // Remove the <script src="packages/test/dart.js"></script> if present.
-      document.querySelector('script[src="packages/test/dart.js"]')?.remove();
+  static AssetId _outputFor(BuildStep buildStep) {
+    return new AssetId(buildStep.inputId.package, p.join('test', 'index.html'));
+  }
 
-      debugHtml = document.outerHtml;
-    } else {
-      debugHtml = '''
-<html>
-  <head>
-    <script src="$jsScriptPath"></script>
-  </head>
-</html>
-      ''';
+  static String _generateHtml(Iterable<AssetId> tests) {
+    if (tests.isEmpty) {
+      return '<strong>No tests found!</strong>';
+    }
+    final buffer = new StringBuffer('    <ul>');
+    for (final test in tests) {
+      final path = p.joinAll(p.split(_debugHtmlId(test).path)..removeAt(0));
+      buffer.writeln('      <li><a href="/$path">${test.path}</a></li>');
     }
-    await buildStep.writeAsString(
-        buildStep.inputId.changeExtension('.debug.html'), debugHtml);
+    buffer.writeln('    </ul>');
+    return buffer.toString();
+  }
+
+  const DebugIndexBuilder();
+
+  @override
+  final buildExtensions = const {
+    r'$test$': const ['index.html'],
+  };
+
+  @override
+  Future<void> build(BuildStep buildStep) async {
+    final files = await buildStep.findAssets(_allTests).toList();
+    final output = _outputFor(buildStep);
+    return buildStep.writeAsString(
+        output,
+        '<html>\n'
+        '  <body>\n'
+        '    ${_generateHtml(files)}\n'
```

Comment:
Add FYI here to open your developer console to see test results?

Reply:
Lets do that in a follow-up to keep this CL smaller.